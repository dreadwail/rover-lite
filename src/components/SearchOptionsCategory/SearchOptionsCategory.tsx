import React, { PureComponent } from 'react';

import { OptionCategory } from '../../state/searchOptions/types';

export interface SearchOptionsCategoryProps {
  readonly categoryId: OptionCategory['categoryId'];
  readonly displayName: OptionCategory['displayName'];
}

export interface SearchOptionsCategoryState {
  readonly isExpanded: boolean;
}

export default class SearchOptionsCategory extends PureComponent<
  SearchOptionsCategoryProps,
  SearchOptionsCategoryState
> {
  constructor(props: SearchOptionsCategoryProps) {
    super(props);

    this.state = { isExpanded: props.categoryId === 'detail' };
  }

  public render() {
    const { displayName, children } = this.props;
    const { isExpanded } = this.state;
    const collapseClassName = isExpanded ? 'collapse.show' : 'collapse';
    const collapseArrow = isExpanded ? '▼' : '▶';
    return (
      <>
        <span className="form-text border-bottom text-secondary mb-2">
          <button
            type="button"
            className="btn btn-link pl-0"
            style={{ textDecoration: 'none' }}
            onClick={this.toggleExpansion}
          >
            <small className="text-muted">{collapseArrow}</small>{' '}
            <span className="text-secondary">{displayName}</span>
          </button>
        </span>
        <div className={collapseClassName}>{children}</div>
      </>
    );
  }

  private readonly toggleExpansion = () => {
    this.setState(state => ({ isExpanded: !state.isExpanded }));
  };
}
