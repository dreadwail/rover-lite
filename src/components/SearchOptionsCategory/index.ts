export * from './SearchOptionsCategory';
export { default } from './SearchOptionsCategory';
