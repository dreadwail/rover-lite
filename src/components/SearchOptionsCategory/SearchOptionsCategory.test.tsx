import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

import SearchOptionsCategory, {
  SearchOptionsCategoryProps,
  SearchOptionsCategoryState,
} from './SearchOptionsCategory';

describe('SearchOptionsCategory', () => {
  let props: SearchOptionsCategoryProps;
  let rendered: ShallowWrapper<SearchOptionsCategoryProps, SearchOptionsCategoryState>;

  beforeEach(() => {
    props = {
      categoryId: 'test-category-id',
      displayName: 'Test Category',
    };
  });

  describe('for the detail category', () => {
    beforeEach(() => {
      props = {
        categoryId: 'detail',
        displayName: 'Detail',
      };
      rendered = shallow(<SearchOptionsCategory {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('for a non-detail category', () => {
    beforeEach(() => {
      props = {
        categoryId: 'test-category-id',
        displayName: 'Test Category',
      };
      rendered = shallow(<SearchOptionsCategory {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('when collapsed', () => {
    beforeEach(() => {
      rendered = shallow(<SearchOptionsCategory {...props} />);
      rendered.setState({ isExpanded: false });
    });

    describe('when clicking to expand', () => {
      beforeEach(() => {
        rendered.find('button').simulate('click');
      });

      it('expands the category', () => {
        expect(rendered).toMatchSnapshot();
      });
    });
  });

  describe('when expanded', () => {
    beforeEach(() => {
      rendered = shallow(<SearchOptionsCategory {...props} />);
      rendered.setState({ isExpanded: true });
    });

    describe('when clicking to collapse', () => {
      beforeEach(() => {
        rendered.find('button').simulate('click');
      });

      it('collapses the category', () => {
        expect(rendered).toMatchSnapshot();
      });
    });
  });
});
