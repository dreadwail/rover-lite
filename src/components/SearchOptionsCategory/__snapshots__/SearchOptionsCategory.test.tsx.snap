// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchOptionsCategory for a non-detail category renders correctly 1`] = `
<Fragment>
  <span
    className="form-text border-bottom text-secondary mb-2"
  >
    <button
      className="btn btn-link pl-0"
      onClick={[Function]}
      style={
        Object {
          "textDecoration": "none",
        }
      }
      type="button"
    >
      <small
        className="text-muted"
      >
        ▶
      </small>
       
      <span
        className="text-secondary"
      >
        Test Category
      </span>
    </button>
  </span>
  <div
    className="collapse"
  />
</Fragment>
`;

exports[`SearchOptionsCategory for the detail category renders correctly 1`] = `
<Fragment>
  <span
    className="form-text border-bottom text-secondary mb-2"
  >
    <button
      className="btn btn-link pl-0"
      onClick={[Function]}
      style={
        Object {
          "textDecoration": "none",
        }
      }
      type="button"
    >
      <small
        className="text-muted"
      >
        ▼
      </small>
       
      <span
        className="text-secondary"
      >
        Detail
      </span>
    </button>
  </span>
  <div
    className="collapse.show"
  />
</Fragment>
`;

exports[`SearchOptionsCategory when collapsed when clicking to expand expands the category 1`] = `
<Fragment>
  <span
    className="form-text border-bottom text-secondary mb-2"
  >
    <button
      className="btn btn-link pl-0"
      onClick={[Function]}
      style={
        Object {
          "textDecoration": "none",
        }
      }
      type="button"
    >
      <small
        className="text-muted"
      >
        ▼
      </small>
       
      <span
        className="text-secondary"
      >
        Test Category
      </span>
    </button>
  </span>
  <div
    className="collapse.show"
  />
</Fragment>
`;

exports[`SearchOptionsCategory when expanded when clicking to collapse collapses the category 1`] = `
<Fragment>
  <span
    className="form-text border-bottom text-secondary mb-2"
  >
    <button
      className="btn btn-link pl-0"
      onClick={[Function]}
      style={
        Object {
          "textDecoration": "none",
        }
      }
      type="button"
    >
      <small
        className="text-muted"
      >
        ▶
      </small>
       
      <span
        className="text-secondary"
      >
        Test Category
      </span>
    </button>
  </span>
  <div
    className="collapse"
  />
</Fragment>
`;
